File: software_design_and_engineering/Original_java/Task.java
import java.util.Comparator;

public class Task {

    private String taskId;
    private String taskName;
    private String taskDescription;

    public Task(String taskId, String taskName, String taskDescription) {
        // Setting the values for the constructor using the set methods (validations are there already)
        this.setTaskId(taskId);
        this.setTaskName(taskName);
        this.setTaskDescription(taskDescription);
    }

    public String getTaskId() {
        return taskId;
    }

    private void setTaskId(String taskId) {
        // Since I set this method to private, then it should not be updatable outside of the class
        // And I will only call it inside the constructor
        if (taskId != null && taskId.length() < 11) {
            this.taskId = taskId;
        } else {
            throw new IllegalArgumentException("Invalid Task Id");
        }
    }

    public String getTaskName() {
        return taskName;
    }

    public void setTaskName(String taskName) {
        if (taskName != null && taskName.length() < 21) {
            this.taskName = taskName;
        } else {
            throw new IllegalArgumentException("Invalid Task Name");
        }
    }

    public String getTaskDescription() {
        return taskDescription;
    }

    public void setTaskDescription(String taskDescription) {
        if (taskDescription != null && taskDescription.length() < 51) {
            this.taskDescription = taskDescription;
        } else {
            throw new IllegalArgumentException("Invalid Task Description");
        }
    }

    // Found this example in stack overflow/(mr. google), this seems a more thorough way to compare the task objects
    public static Comparator<Task> compareById = new Comparator<Task>() {
        @Override
        public int compare(Task t1, Task t2) {
            return t1.getTaskId().compareTo(t2.getTaskId());
        }
    };

    // Overriding the equals operator to compare (extension of the stack overflow example)
    // A simpiler way I first did was just to compare the taskIds
    @Override
    public boolean equals(Object obj) {
        if (this == obj)
            return true;

        if (obj == null)
            return false;

        if (this.getClass() != obj.getClass())
            return false;

        Task t = (Task) obj;
        return getTaskId().equals(t.getTaskId());
    }

    // Instead of printing the string for testing, I found we can override the toString method
    @Override
    public String toString() {
        return "taskId: " + getTaskId() +
                "\ntaskName: " + getTaskName() +
                "\ntaskDescription: " + getTaskDescription() + "\n";
    }

}

File: software_design_and_engineering/Enhanced_python/task.py
class Task:
    task_id = None
    task_name = None
    task_description = None

    def __init__(self, task_id, task_name, task_description):
        # Setting the values for the constructor using the set methods (validations are there already)
        self.set_task_id(task_id)
        self.set_task_name(task_name)
        self.set_task_description(task_description)

    def __eq__(self, other):
        # Defining the __eq__ so we can compare two task objects by task_id
        if (isinstance(other, C)):
            return self.task_id == other.task_id
        return false

    def __str__(self):
        # Definig the __str__ method so we can print the class variables
        return f"task_id: {task_id}\ntask_name: {task_name}\ntask_description: {task_description}"

    def set_task_id(self, task_id):
        if task_id is not None and len(task_id) < 11:
            self.task_id = task_id
        else:
            raise IllegalArgumentException("Invalid Task Id")

    def set_task_name(self, task_name):
        if task_name is not None and len(task_name) < 21:
            self.task_name = task_name
        else:
            raise IllegalArgumentException("Invalid Task Name")

    def set_task_description(self, task_description):
        if task_description is not None and len(task_description) < 51:
            self.task_description = task_description
        else:
            raise IllegalArgumentException("Invalid Task Description")

    def get_task_id(self):
        return self.task_id

    def get_task_name(self):
        return self.task_name

    def get_task_description(self):
        return self.task_description


class IllegalArgumentException(Exception):
    def __init__(self, message):
        print(f'ERROR: {message}')
